source/simulator/solver_schemes.cc

```diff
@@ -1452,6 +1455,149 @@ namespace aspect
+
+    // determine how often to iterate over the forward & adjoint stokes equations
+    const unsigned int max_nonlinear_iterations =
+      (pre_refinement_step < parameters.initial_adaptive_refinement)
+      ?
+      std::min(parameters.max_nonlinear_iterations,
+               parameters.max_nonlinear_iterations_in_prerefinement)
+      :
+      parameters.max_nonlinear_iterations;
+
+    SolverControl nonlinear_solver_control(max_nonlinear_iterations,
+                                           parameters.nonlinear_tolerance);
+
+    // follow the standardized format for nonlinear solver loops. We are not currently recomputing
+    // a residual in the iteration, which means that the number of iterations is always given by the
+    // maximum number of iterations.
+    double relative_residual = std::numeric_limits<double>::max();
+
+    // start the iteration.
+    nonlinear_iteration = 0;
+    do
+      {
+        pcout << "      Number of current adjoint iteration: " << nonlinear_iteration+1  << std::endl;
+        pcout << "      Solving the forward problem ... " << std::endl;
+
+        // set the assemblers for the RHS
+        set_assemblers();
+
+        // build the stokes system
+        rebuild_stokes_matrix = rebuild_stokes_preconditioner = true;
+
+        assemble_stokes_system();
+        build_stokes_preconditioner();
+
+        // solve the stokes system and put the solution (x in Ax=b) into 'solution'
+        solve_stokes();
+
+        // save the solution from overwriting during the adjoint problem by
+        // storing it in 'current_linearization_point'
+        current_linearization_point.block(introspection.block_indices.velocities)
+          = solution.block(introspection.block_indices.velocities);
+        if (introspection.block_indices.velocities != introspection.block_indices.pressure)
+          current_linearization_point.block(introspection.block_indices.pressure)
+            = solution.block(introspection.block_indices.pressure);
+
+        // calculate the dynamic topography based on the forward solution,
+        // which will be needed for the new RHS assembly of the adjoint system
+        Postprocess::DynamicTopography<dim> &dynamic_topography = const_cast<Postprocess::DynamicTopography<dim> &> (
+                                                                    postprocess_manager.template get_matching_postprocessor<Postprocess::DynamicTopography<dim> >());
+
+        dynamic_topography.execute(statistics);
+
+        // save the global statics output of the forward solution
+        Postprocess::GlobalStatistics<dim> &global_statistics = const_cast<Postprocess::GlobalStatistics<dim> &> (
+                                                                  postprocess_manager.template get_matching_postprocessor<Postprocess::GlobalStatistics<dim> >());
+
+        global_statistics.execute(statistics);
```

Comment:
I don't understand the order of things here. You first save the state and then compute statistics. What do you actually do with the saved statistics?

Reply:
I thought it could be useful to have the statistics for the forward run, but maybe that's not necessary (and more confusing than helpful). I'll remove this.